common/achievement/get.ts

```diff
@@ -0,0 +1,214 @@
+const getFurtherAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID, template: { isActive: true } },
+        include: { template: true },
+    });
+
+    const groups = Array.from(new Set(userAchievements.map((ua) => ua.template.group)));
+    const templates = Array.from(new Set(userAchievements.map((ua) => ua.templateId)));
+    const tieredTemplates = await prisma.achievement_template.findMany({
+        where: {
+            isActive: true,
+            group: { in: groups },
+            type: achievement_type_enum.TIERED,
+            NOT: { id: { in: templates } },
+        },
+    });
+
+    const tieredAchievements = tieredTemplates.map((template) => {
+        const dataAggregationKeys = Object.keys(template.conditionDataAggregations);
+        const maxValue = dataAggregationKeys
+            .map((key) => {
+                return Number(template.conditionDataAggregations[key].valueToAchieve);
+            })
+            .reduce((a, b) => a + b, 0);
+        return {
+            id: template.id,
+            name: template.name,
+            subtitle: template.subtitle,
+            description: template.description,
+            image: getAchievementImageURL(template.image),
+            alternativeText: 'alternativeText',
+            actionType: template.actionType as achievement_action_type_enum,
+            achievementType: template.type as achievement_type_enum,
+            achievementState: achievement_state.INACTIVE,
+            steps: null,
+            maxSteps: maxValue,
+            currentStep: 0,
+            isNewAchievement: null,
+            progressDescription: `Noch ${userAchievements.length - userAchievements.length} Schritte bis zum Abschluss`,
+            actionName: template.actionName,
+            actionRedirectLink: template.actionRedirectLink,
+        };
+    });
+    return tieredAchievements;
+};
+// User achievements are already started by the user and are either active or completed.
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID, AND: { template: { isActive: true } } },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = await generateReorderedAchievementData(userAchievementGroups, user);
+    return achievements;
+};
+
+const generateReorderedAchievementData = async (groups: { [group: string]: User_achievement[] }, user: User): Promise<Achievement[]> => {
```

Comment:
I think the functions of this file should use the prisma instead of the graphql type as all the data is coming from the DB.

Reply:
ok everything was retyped but it now was necessary including a new type in the file to secure right type of the achievement group including templates.